File: src/cpp_pubsub/src/topics_practice.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <turtlesim/msg/pose.hpp>

using namespace std::chrono_literals;

class TopicsPractice : public rclcpp::Node
{
public:
    TopicsPractice()
    : Node("topics_practice"), countdown_(5), state_("countdown"), spiral_size_(0.5)
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtle1/cmd_vel", 10);
        timer_ = this->create_wall_timer(500ms, std::bind(&TopicsPractice::timer_callback, this));
        pose_subscriber_ = this->create_subscription<turtlesim::msg::Pose>(
            "/turtle1/pose", 10, std::bind(&TopicsPractice::pose_callback, this, std::placeholders::_1));
    }

private:
    void pose_callback(const turtlesim::msg::Pose::SharedPtr msg)
    {
        pose_ = msg;
    }

    void timer_callback()
    {
        if (state_ == "countdown")
        {
            RCLCPP_INFO(this->get_logger(), "Countdown: %d", countdown_);
            countdown_--;
            if (countdown_ < 0)
            {
                state_ = "drawing_spiral";
                RCLCPP_INFO(this->get_logger(), "Drawing spiral");
            }
        }
        else if (state_ == "drawing_spiral")
        {
            RCLCPP_INFO(this->get_logger(), "Drawing spiral");
            auto msg = geometry_msgs::msg::Twist();
            msg.linear.x = spiral_size_;
            msg.angular.z = 1.5;
            spiral_size_ += 0.05;
            publisher_->publish(msg);

            if (pose_ && (pose_->x < 1.0 || pose_->x > 7.0 || pose_->y < 1.0 || pose_->y > 7.0))
            {
                state_ = "going_straight";
                RCLCPP_INFO(this->get_logger(), "Going straight");
            }
        }
        else if (state_ == "going_straight")
        {
            RCLCPP_INFO(this->get_logger(), "Going straight");
            auto msg = geometry_msgs::msg::Twist();
            msg.linear.x = 2.0;
            msg.angular.z = 0.0;
            publisher_->publish(msg);
        }
    }

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_subscriber_;
    rclcpp::TimerBase::SharedPtr timer_;
    turtlesim::msg::Pose::SharedPtr pose_;
    int countdown_;
    std::string state_;
    double spiral_size_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TopicsPractice>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/py_pubsub/py_pubsub/topics_practice.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
import time

class TopicsPractice(Node):

    def __init__(self):
        super().__init__('topics_practice')
        self.publisher_ = self.create_publisher(Twist, '/turtle1/cmd_vel', 10)
        self.timer = self.create_timer(0.5, self.timer_callback)
        self.countdown = 5
        self.state = 'countdown'
        self.pose_subscriber = self.create_subscription(Pose, '/turtle1/pose', self.pose_callback, 10)
        self.pose = None
        
    def pose_callback(self, msg):
        self.pose = msg    

    def timer_callback(self):
        if self.state == 'countdown':
            self.get_logger().info(f'Countdown: {self.countdown}')
            self.countdown -= 1
            if self.countdown < 0:
                self.state = 'drawing_spiral'
                self.get_logger().info('Drawing spiral')
                self.spiral_size = 0.5

        elif self.state == 'drawing_spiral':
            self.get_logger().info('Drawing spiral')
            msg = Twist()
            msg.linear.x = self.spiral_size
            msg.angular.z = 1.5
            self.spiral_size += 0.05
            self.publisher_.publish(msg)

            if self.pose and (self.pose.x < 1.0 or self.pose.x > 7.0 or self.pose.y < 1.0 or self.pose.y > 7.0):
                self.state = 'going_straight'
                self.get_logger().info('Going straight')

        elif self.state == 'going_straight':
            self.get_logger().info('Going straight')
            msg = Twist()
            msg.linear.x = 2.0
            msg.angular.z = 0.0
            self.publisher_.publish(msg)

def main(args=None):
    rclpy.init(args=args)
    node = TopicsPractice()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
